### PR TITLE
chore: Refactor TokenProviderExtension

### PR DESCRIPTION
**Pull request by CodeBlanch**
Relates to #3901

# Changes

* Refactor reusable bits of `TokenProviderExtension` into `BackgroundProviderExtension`

# Details

I'm working on adding `ApiKey` and `BasicAuth` capabilities and there will be flat file extensions implemented initially. There is code in `TokenProviderExtension` which handles the background refresh of `BearerToken`s. What this PR does is refactor the common things so refresh logic can be reused for any generic struct (`BearerToken`, `ApiKey`, `BasicAuthCredentials`, etc.).

### CODE REVIEW

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs
Comment (utpilla) [inline rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs]: Treating None as “refresh in one year” works for the existing bearer-token providers, but it may not support the API-key and Basic-auth providers motivating this extraction. File-backed credentials commonly have no expiry while still requiring periodic polling to detect rotation. Once cached, such a value would remain fresh and its source would not be checked again for roughly a year.

Could the source define its next-refresh policy, perhaps through a refresh interval or `next_refresh` method? Expiry-based providers could retain the current behavior, while file-backed providers could supply their polling interval.

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs [hunk 1]
```diff
@@ -135,37 +150,42 @@ impl<S: TokenSource, M: TokenProviderMetrics> TokenProviderExtension<S, M> {
...
+    #[must_use]
+    pub fn current_fresh_value(&self) -> Option<T> {
+        // The value lives inside the watch channel behind a temporary read
         // guard; clone it out so we can return an owned value (and release the
-        // guard, which would otherwise block the writer). `BearerToken` clones
+        // guard, which would otherwise block the writer). `T` clones
         // are cheap: a refcount bump on the shared secret.
-        let token = self.tx.borrow().clone()?;
-        match token.expires_on() {
+        let value = self.inner.tx.borrow().clone()?;
```
Comment (utpilla) [inline rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs]: The comment assumes cloning `T` is always a refcount bump, but the generic contract only requires `T: Clone`. This method runs on every cache hit, so a future value type could unintentionally introduce a deeper clone on the request path.

Could we document that `T` must be cheaply cloneable?

File: rust/otap-dataflow/crates/contrib-extensions/src/common/token_refresh/extension.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        }
+        // Negative cache: if the most recent acquisition failed within the
+        // cooldown window, surface the throttle instead of hitting the token
+        // endpoint again. The background loop keeps retrying on its own cadence.
+        if self.recently_failed() {
+            return Err(self.capability_error("token acquisition throttled after recent failure"));
+        }
+        self.refresh_once()
+            .await
+            .map_err(|err| self.capability_error(err))
```
Comment (utpilla) [inline rust/otap-dataflow/crates/contrib-extensions/src/common/token_refresh/extension.rs]: This adapter still assembles the cache fast path, fetch lock, double-check, negative cache, refresh, and error conversion from the generic provider’s low-level methods. Future API-key and Basic-auth adapters will need to reproduce the same synchronization-sensitive sequence.

Could `BackgroundProviderExtension` expose a generic `get_value()` method that owns this algorithm? Capability adapters would then only provide capability-specific error wording and stream adaptation, making the abstraction harder to use incorrectly.

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs
Comment (CodeBlanch) [reply rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs]: Updated

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs (hunk 1 above)
Comment (CodeBlanch) [reply rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs]: Updated

File: rust/otap-dataflow/crates/contrib-extensions/src/common/token_refresh/extension.rs (hunk 2 above)
Comment (CodeBlanch) [reply rust/otap-dataflow/crates/contrib-extensions/src/common/token_refresh/extension.rs]: I refactored this logic into a `get_value` which had the nice benefit of allowing me to make some of the other methods private or remove them completely. Nice!

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs [hunk 3]
```diff
@@ -230,47 +289,51 @@ impl<S: TokenSource, M: TokenProviderMetrics> Inner<S, M> {
...
+    let min_next = now + Duration::from_secs(MIN_REFRESH_INTERVAL_SECS);
+    match value_expires_on {
         Some(expires_on) => {
             let target = tokio::time::Instant::from_std(expires_on)
                 .checked_sub(expiry_buffer)
                 .unwrap_or(now);
             target.max(min_next)
         }
-        None => now + Duration::from_secs(NON_EXPIRING_REFRESH_SECS),
+        None => now + non_expiring_refresh_interval,
```
Comment (utpilla) [inline rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs:334]: This branch does not apply the minimum refresh interval. A non-expiring provider configured with `Duration::ZERO` could continuously fetch and publish, while any interval below 10 seconds would bypass the documented floor.

It would also be useful to test zero and sub-10-second intervals.

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs [hunk 4]
```diff
@@ -93,40 +97,55 @@ struct FailureState {
...
+    BackgroundProviderExtension<S, M, T, C>
+{
     /// Builds a new extension instance.
     #[must_use]
     pub fn new(
         name: &str,
         source: S,
+        usable_margin: Duration,
+        non_expiring_refresh_interval: Duration,
         expiry_buffer: Duration,
```
Comment (utpilla) [inline rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs]: nit: A small `BackgroundRefreshPolicy` struct make this clearer and provide one place to validate these settings:

```rust
struct BackgroundRefreshPolicy {
    usable_margin: Duration,
    non_expiring_interval: Duration,
    expiry_buffer: Duration,
}
```

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/metrics.rs [hunk 5]
```diff
@@ -1,42 +1,37 @@
...
-pub struct TokenProviderMetricsTracker<M: TokenProviderMetrics> {
+/// Tracks and flushes a background provider's metric set.
+pub struct BackgroundProviderMetricsTracker<M: BackgroundProviderMetrics> {
     metrics: MetricSet<M>,
 }
 
-impl<M: TokenProviderMetrics> std::fmt::Debug for TokenProviderMetricsTracker<M> {
+impl<M: BackgroundProviderMetrics> std::fmt::Debug for BackgroundProviderMetricsTracker<M> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         f.debug_struct("TokenProviderMetricsTracker").finish()
```
Comment (utpilla) [inline rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/metrics.rs]:
```suggestion
        f.debug_struct("BackgroundProviderMetricsTracker").finish()
```

File: rust/otap-dataflow/crates/contrib-extensions/src/common/token_refresh/mod.rs
Comment (utpilla) [inline rust/otap-dataflow/crates/contrib-extensions/src/common/token_refresh/mod.rs]: Let's update the doc comments to the new names: `BackgroundProviderSource` and `BackgroundProviderMetrics`.

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/tests.rs [hunk 6]
```diff
@@ -13,6 +13,10 @@ use super::provider::{
 
 // -- schedule_next timing tests --------------------------------
 
+/// Next-refresh delay used for non-expiring values (~1 year). The loop is still
+/// woken by control messages in the meantime.
+const NON_EXPIRING_REFRESH_SECS: u64 = 365 * 24 * 60 * 60;
+
 // Scenario: Schedule the next refresh for a token expiring in ~1 hour with a 5m buffer.
```
Comment (utpilla) [inline rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/tests.rs]: Tests in this repository seem to use doc comments (`///`) instead of plain `//` comments for `Scenario` and `Guarantees`. Let's use `///` to be consistent.

**Review by utpilla — APPROVED** [review]
Left some suggestions. Looks good overall.

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs (hunk 3 above)
Comment (CodeBlanch) [reply rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs:334]: See below.

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs (hunk 4 above)
Comment (CodeBlanch) [reply rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs]: I added the struct and put some validation in there to make sure we use max of `non_expiring_interval` or `MIN_REFRESH_INTERVAL_SECS`.

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/metrics.rs (hunk 5 above)
Comment (CodeBlanch) [reply rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/metrics.rs]: I made this change but it is breaking a couple tests. Pinged @pritishnahar95 to confirm whether or not the name is actually used in telemetry. I'll do _something_ to fix it based on what he says.

File: rust/otap-dataflow/crates/contrib-extensions/src/azure_identity_auth/tests.rs [hunk 7]
```diff
@@ -20,7 +20,7 @@ use super::config::{AuthMethod, Config};
 use super::error::Error;
 use super::metrics::AzureIdentityAuthMetrics;
 use super::*;
-use crate::common::token_refresh::TokenProviderMetricsTracker;
+use crate::common::background_refresh::BackgroundProviderMetricsTracker;
```
Comment (lquerel) [inline rust/otap-dataflow/crates/contrib-extensions/src/azure_identity_auth/tests.rs]: [P1] Update the debug-name assertions after the tracker rename

This import renames the tracker, but the unchanged assertions in `azure_identity_auth/tests.rs:404` and `oauth2_client_auth/tests.rs:854` still require the old `TokenProviderMetricsTracker` debug name. `cargo test -p otel-arrow-dfe-contrib-extensions --all-features metrics_tracker_records_snapshots_and_reports` fails both tests locally, matching the required Linux and Windows CI failures. Update both assertions for `BackgroundProviderMetricsTracker`, or assert stable behavior instead of a debug type name.

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/metrics.rs (hunk 5 above)
Comment (pritishnahar95) [reply rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/metrics.rs]: I believe the `Debug` struct name is not used as a telemetry identifier. The metric set names are defined independently by `[#[metric_set(name = "...")]]`, and the tracker reports the wrapped `MetricSet` directly.

My guess is that the failing tests are only asserting the old debug output `TokenProviderMetricsTracker`. Since the type was renamed, changing the label to `BackgroundProviderMetricsTracker` is correct, and those assertions should be updated accordingly. This only changes human-readable debug formatting, not emitted metric names or attributes.

**Review by drewrelmas — APPROVED** [review]

**Review by pritishnahar95 — APPROVED** [review]

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs [hunk 8]
```diff
@@ -93,40 +97,77 @@ struct FailureState {
...
+        usable_margin: Duration,
+        mut non_expiring_refresh_interval: Duration,
+        expiry_buffer: Duration,
+    ) -> BackgroundProviderRefreshPolicy {
+        // Note: We don't allow zero duration for non_expiring_refresh_interval
+        // because that would cause a busy loop in refresh.
+        non_expiring_refresh_interval =
+            non_expiring_refresh_interval.max(Duration::from_secs(MIN_REFRESH_INTERVAL_SECS));
+
+        Self {
```
Comment (gouslu) [inline rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs]: Could this constructor reject `expiry_buffer <= usable_margin`? `schedule_next` refreshes at `expires_on - expiry_buffer`, while consumers stop using the value at `expires_on - usable_margin`. If the buffer is smaller, refresh is deliberately scheduled after the value becomes unusable; if they are equal, acquisition latency still creates an availability gap.

OAuth validates this separately and Azure currently supplies safe constants, but this generic policy is intended for future API-key and Basic-auth providers. Leaving the invariant to every caller makes the reusable abstraction easy to misuse. Could policy construction be fallible and enforce `expiry_buffer > usable_margin`, with a regression test?

File: rust/otap-dataflow/crates/contrib-extensions/src/common/mod.rs [hunk 9]
```diff
@@ -3,6 +3,12 @@
 
 //! Shared functions and data types for contrib extension implementations.
 
+#[cfg(any(
+    feature = "azure-identity-auth-extension",
+    feature = "oauth2-client-auth-extension"
+))]
+pub mod background_refresh;
```
Comment (gouslu) [inline rust/otap-dataflow/crates/contrib-extensions/src/common/mod.rs:10]: Is this intended to become part of the crate’s public API? Since `common` is public, `pub mod background_refresh` exposes the new extension, policy, source, and metrics types externally. To me, it seems like this should remain crate-private (`pub(crate)`) within `contrib-extensions`.

**Review by jmacd — APPROVED** [review]

File: rust/otap-dataflow/crates/contrib-extensions/src/common/mod.rs (hunk 9 above)
Comment (CodeBlanch) [reply rust/otap-dataflow/crates/contrib-extensions/src/common/mod.rs:10]: `token_refresh` was public before so this really just went along with that. But I do agree with you we shouldn't expose the background stuff. I did this: [#200da5](https://github.com/open-telemetry/otel-arrow/pull/3979/commits/2000da595e4413f4b9553448a736546d0242cd00)

**Review by gouslu — APPROVED** [review]
Approving with the expectation that https://github.com/open-telemetry/otel-arrow/pull/3979#discussion_r3962463101 will be addressed. I don’t think the policy should allow `expiry_buffer <= usable_margin`, since that can schedule a refresh after the credential is already considered unusable.

File: rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs (hunk 8 above)
Comment (CodeBlanch) [reply rust/otap-dataflow/crates/contrib-extensions/src/common/background_refresh/provider.rs]: I pushed this change: https://github.com/open-telemetry/otel-arrow/pull/3979/commits/7a3c7ca65a0dd7b1dc0bfc31714999b36a9d8189

LMK if that works.
